File: OpenBurnBarCore/Sources/OpenBurnBarMedia/MediaFrameV2.swift
import Foundation

public struct MediaFrameV2: Sendable, Equatable {
    public struct Kind: RawRepresentable, Sendable, Codable, Equatable, Hashable {
        public var rawValue: UInt8

        public init(rawValue: UInt8) {
            self.rawValue = rawValue
        }

        public static let videoNAL = Kind(rawValue: 0x01)
        public static let audioOpus = Kind(rawValue: 0x02)
        public static let videoDatagram = Kind(rawValue: 0x03)
        public static let rtcSenderReport = Kind(rawValue: 0x11)
        public static let rtcReceiverReport = Kind(rawValue: 0x12)
        public static let rpsRecoveryRequest = Kind(rawValue: 0x13)
        public static let codecNegotiation = Kind(rawValue: 0x14)
    }

    public var kind: Kind
    public var flags: UInt16
    public var gopID: UInt32
    public var frameIndex: UInt32
    public var presentationTimestampMillis: UInt64
    public var metadata: Data
    public var payload: Data

    public init(
        kind: Kind,
        flags: UInt16 = 0,
        gopID: UInt32 = 0,
        frameIndex: UInt32 = 0,
        presentationTimestampMillis: UInt64 = 0,
        metadata: Data = Data(),
        payload: Data = Data()
    ) {
        self.kind = kind
        self.flags = flags
        self.gopID = gopID
        self.frameIndex = frameIndex
        self.presentationTimestampMillis = presentationTimestampMillis
        self.metadata = metadata
        self.payload = payload
    }
}

public struct MediaFrameV2Metadata: Codable, Sendable, Equatable {
    /// Codec used for this frame. This is descriptive metadata for receivers
    /// and benchmark traces; codec selection still happens during negotiation.
    public var codec: MercuryVideoCodec?
    /// VideoToolbox LTR acknowledgement token returned on the encoded sample.
    /// Receivers ACK this value only after successful receipt/decode.
    public var longTermReferenceToken: MercuryLTRToken?

    public init(
        codec: MercuryVideoCodec? = nil,
        longTermReferenceToken: MercuryLTRToken? = nil
    ) {
        self.codec = codec
        self.longTermReferenceToken = longTermReferenceToken
    }

    public func encode() throws -> Data {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys]
        return try encoder.encode(self)
    }

    public static func decode(_ data: Data) throws -> MediaFrameV2Metadata {
        guard !data.isEmpty else { return MediaFrameV2Metadata() }
        return try JSONDecoder().decode(MediaFrameV2Metadata.self, from: data)
    }
}

public struct MediaFrameV2Codec: Sendable {
    public enum CodecError: Error, Equatable, Sendable {
        case notNegotiated
        case envelopeTooShort
        case invalidMagic
        case unsupportedVersion(UInt8)
        case payloadTooLarge(actual: Int, max: Int)
        case headerTruncated
    }

    public static let magic = Data([0x4F, 0x42, 0x4D, 0x46, 0x56, 0x32]) // OBMFV2
    public static let version: UInt8 = 2
    public static let fixedHeaderByteCount = 34
    public static let defaultMaxPayloadBytes: Int = 256 * 1024

    public let maxPayloadBytes: Int

    public init(maxPayloadBytes: Int = MediaFrameV2Codec.defaultMaxPayloadBytes) {
        self.maxPayloadBytes = maxPayloadBytes
    }

    public static func isEncodedEnvelope(_ envelope: Data) -> Bool {
        let headerStart = 4
        guard envelope.count >= headerStart + Self.magic.count else { return false }
        let normalized: Data = envelope.withUnsafeBytes { Data($0) }
        return normalized.subdata(in: headerStart..<(headerStart + Self.magic.count)) == Self.magic
    }

    public func encode(
        _ frame: MediaFrameV2,
        negotiatedVersion: MercuryMediaFrameWireVersion
    ) throws -> Data {
        guard negotiatedVersion == .v2 else { throw CodecError.notNegotiated }
        let totalPayloadCount = Self.fixedHeaderByteCount + frame.metadata.count + frame.payload.count
        guard totalPayloadCount <= maxPayloadBytes else {
            throw CodecError.payloadTooLarge(actual: totalPayloadCount, max: maxPayloadBytes)
        }

        var envelope = Data(capacity: 4 + totalPayloadCount)
        appendUInt32BigEndian(UInt32(totalPayloadCount), to: &envelope)
        envelope.append(Self.magic)
        envelope.append(Self.version)
        envelope.append(frame.kind.rawValue)
        appendUInt16BigEndian(frame.flags, to: &envelope)
        appendUInt32BigEndian(frame.gopID, to: &envelope)
        appendUInt32BigEndian(frame.frameIndex, to: &envelope)
        appendUInt64BigEndian(frame.presentationTimestampMillis, to: &envelope)
        appendUInt32BigEndian(UInt32(frame.metadata.count), to: &envelope)
        appendUInt32BigEndian(UInt32(frame.payload.count), to: &envelope)
        envelope.append(frame.metadata)
        envelope.append(frame.payload)
        return envelope
    }

    public func decode(_ envelope: Data) throws -> (frame: MediaFrameV2, consumed: Int) {
        let lengthPrefixBytes = 4
        guard envelope.count >= lengthPrefixBytes + Self.fixedHeaderByteCount else {
            throw CodecError.envelopeTooShort
        }

        let normalized: Data = envelope.withUnsafeBytes { Data($0) }
        let totalPayloadCount = Int(readUInt32BigEndian(from: normalized, at: 0))
        guard totalPayloadCount <= maxPayloadBytes else {
            throw CodecError.payloadTooLarge(actual: totalPayloadCount, max: maxPayloadBytes)
        }
        let totalEnvelopeBytes = lengthPrefixBytes + totalPayloadCount
        guard normalized.count >= totalEnvelopeBytes else {
            throw CodecError.headerTruncated
        }

        let headerStart = lengthPrefixBytes
        guard normalized.subdata(in: headerStart..<(headerStart + Self.magic.count)) == Self.magic else {
            throw CodecError.invalidMagic
        }
        let version = normalized[headerStart + Self.magic.count]
        guard version == Self.version else { throw CodecError.unsupportedVersion(version) }

        let kindOffset = headerStart + Self.magic.count + 1
        let kind = MediaFrameV2.Kind(rawValue: normalized[kindOffset])
        let flags = readUInt16BigEndian(from: normalized, at: kindOffset + 1)
        let gopID = readUInt32BigEndian(from: normalized, at: kindOffset + 3)
        let frameIndex = readUInt32BigEndian(from: normalized, at: kindOffset + 7)
        let pts = readUInt64BigEndian(from: normalized, at: kindOffset + 11)
        let metadataCount = Int(readUInt32BigEndian(from: normalized, at: kindOffset + 19))
        let payloadCount = Int(readUInt32BigEndian(from: normalized, at: kindOffset + 23))
        let metadataStart = headerStart + Self.fixedHeaderByteCount
        let metadataEnd = metadataStart + metadataCount
        let payloadEnd = metadataEnd + payloadCount
        guard payloadEnd <= totalEnvelopeBytes else {
            throw CodecError.headerTruncated
        }

        return (
            MediaFrameV2(
                kind: kind,
                flags: flags,
                gopID: gopID,
                frameIndex: frameIndex,
                presentationTimestampMillis: pts,
                metadata: normalized.subdata(in: metadataStart..<metadataEnd),
                payload: normalized.subdata(in: metadataEnd..<payloadEnd)
            ),
            totalEnvelopeBytes
        )
    }

    private func appendUInt16BigEndian(_ value: UInt16, to data: inout Data) {
        var be = value.bigEndian
        withUnsafeBytes(of: &be) { data.append(contentsOf: $0) }
    }

    private func appendUInt32BigEndian(_ value: UInt32, to data: inout Data) {
        var be = value.bigEndian
        withUnsafeBytes(of: &be) { data.append(contentsOf: $0) }
    }

    private func appendUInt64BigEndian(_ value: UInt64, to data: inout Data) {
        var be = value.bigEndian
        withUnsafeBytes(of: &be) { data.append(contentsOf: $0) }
    }

    private func readUInt16BigEndian(from data: Data, at offset: Int) -> UInt16 {
        var raw: UInt16 = 0
        _ = withUnsafeMutableBytes(of: &raw) { dest in
            data.copyBytes(to: dest, from: offset..<(offset + 2))
        }
        return UInt16(bigEndian: raw)
    }

    private func readUInt32BigEndian(from data: Data, at offset: Int) -> UInt32 {
        var raw: UInt32 = 0
        _ = withUnsafeMutableBytes(of: &raw) { dest in
            data.copyBytes(to: dest, from: offset..<(offset + 4))
        }
        return UInt32(bigEndian: raw)
    }

    private func readUInt64BigEndian(from data: Data, at offset: Int) -> UInt64 {
        var raw: UInt64 = 0
        _ = withUnsafeMutableBytes(of: &raw) { dest in
            data.copyBytes(to: dest, from: offset..<(offset + 8))
        }
        return UInt64(bigEndian: raw)
    }
}

File: android/app/src/main/java/com/openburnbar/data/media/MediaFrameV2.kt
package com.openburnbar.data.media

import java.nio.ByteBuffer
import java.nio.ByteOrder
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

private const val VAL_4 = 4

@JvmInline
value class MediaFrameV2Kind(val rawValue: Byte) {
    companion object {
        val VIDEO_NAL = MediaFrameV2Kind(0x01)
        val AUDIO_OPUS = MediaFrameV2Kind(0x02)
        val VIDEO_DATAGRAM = MediaFrameV2Kind(0x03)
        val RTC_SENDER_REPORT = MediaFrameV2Kind(0x11)
        val RTC_RECEIVER_REPORT = MediaFrameV2Kind(0x12)
        val RPS_RECOVERY_REQUEST = MediaFrameV2Kind(0x13)
        val CODEC_NEGOTIATION = MediaFrameV2Kind(0x14)
    }
}

data class MediaFrameV2(
    val kind: MediaFrameV2Kind,
    val flags: UShort = 0u,
    val gopID: UInt = 0u,
    val frameIndex: UInt = 0u,
    val presentationTimestampMillis: ULong = 0uL,
    val metadata: ByteArray = ByteArray(0),
    val payload: ByteArray = ByteArray(0),
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is MediaFrameV2) return false
        return kind == other.kind &&
            flags == other.flags &&
            gopID == other.gopID &&
            frameIndex == other.frameIndex &&
            presentationTimestampMillis == other.presentationTimestampMillis &&
            metadata.contentEquals(other.metadata) &&
            payload.contentEquals(other.payload)
    }

    override fun hashCode(): Int {
        var result = kind.hashCode()
        result = 31 * result + flags.hashCode()
        result = 31 * result + gopID.hashCode()
        result = 31 * result + frameIndex.hashCode()
        result = 31 * result + presentationTimestampMillis.hashCode()
        result = 31 * result + metadata.contentHashCode()
        result = 31 * result + payload.contentHashCode()
        return result
    }
}

@Serializable
data class MediaFrameV2LongTermReferenceToken(
    val value: Long,
)

@Serializable
data class MediaFrameV2Metadata(
    val codec: String? = null,
    val longTermReferenceToken: MediaFrameV2LongTermReferenceToken? = null,
) {
    fun encode(): ByteArray = json.encodeToString(MediaFrameV2Metadata.serializer(), this).encodeToByteArray()

    companion object {
        private val json =
            Json {
                ignoreUnknownKeys = true
                encodeDefaults = false
            }

        fun decode(bytes: ByteArray): MediaFrameV2Metadata {
            if (bytes.isEmpty()) return MediaFrameV2Metadata()
            return json.decodeFromString(MediaFrameV2Metadata.serializer(), bytes.decodeToString())
        }
    }
}

class MediaFrameV2Codec(
    val maxPayloadBytes: Int = DEFAULT_MAX_PAYLOAD_BYTES,
) {
    sealed class CodecError(message: String) : RuntimeException(message) {
        object NotNegotiated : CodecError("MediaFrame v2 was not negotiated")

        object EnvelopeTooShort : CodecError("envelope too short")

        object InvalidMagic : CodecError("invalid MediaFrame v2 magic")

        data class UnsupportedVersion(val version: Byte) : CodecError("unsupported MediaFrame v2 version: $version")

        data class PayloadTooLarge(val actual: Int, val max: Int) : CodecError("payload too large: $actual > $max")

        object HeaderTruncated : CodecError("header truncated")
    }

    data class Decoded(val frame: MediaFrameV2, val consumed: Int)

    fun encode(frame: MediaFrameV2, negotiatedVersion: MercuryMediaFrameWireVersion): ByteArray {
        if (negotiatedVersion != MercuryMediaFrameWireVersion.V2) {
            throw CodecError.NotNegotiated
        }
        val totalPayloadCount = FIXED_HEADER_BYTE_COUNT + frame.metadata.size + frame.payload.size
        if (totalPayloadCount > maxPayloadBytes) {
            throw CodecError.PayloadTooLarge(totalPayloadCount, maxPayloadBytes)
        }
        return ByteBuffer.allocate(VAL_4 + totalPayloadCount)
            .order(ByteOrder.BIG_ENDIAN)
            .putInt(totalPayloadCount)
            .put(MAGIC)
            .put(VERSION)
            .put(frame.kind.rawValue)
            .putShort(frame.flags.toShort())
            .putInt(frame.gopID.toInt())
            .putInt(frame.frameIndex.toInt())
            .putLong(frame.presentationTimestampMillis.toLong())
            .putInt(frame.metadata.size)
            .putInt(frame.payload.size)
            .put(frame.metadata)
            .put(frame.payload)
            .array()
    }

    fun decode(envelope: ByteArray): Decoded {
        val header = decodeEnvelopeHeader(envelope)
        val buffer = header.buffer
        validateMagicAndVersion(buffer)

        val kind = MediaFrameV2Kind(buffer.get())
        val flags = buffer.short.toUShort()
        val gopID = buffer.int.toUInt()
        val frameIndex = buffer.int.toUInt()
        val pts = buffer.long.toULong()
        val metadataLength = buffer.int
        val payloadLength = buffer.int
        validatePayloadBounds(metadataLength, payloadLength, buffer, header.totalEnvelopeBytes)
        val metadata = ByteArray(metadataLength)
        buffer.get(metadata)
        val payload = ByteArray(payloadLength)
        buffer.get(payload)
        return Decoded(
            frame =
            MediaFrameV2(
                kind = kind,
                flags = flags,
                gopID = gopID,
                frameIndex = frameIndex,
                presentationTimestampMillis = pts,
                metadata = metadata,
                payload = payload,
            ),
            consumed = header.totalEnvelopeBytes,
        )
    }

    private data class EnvelopeHeader(
        val buffer: ByteBuffer,
        val totalEnvelopeBytes: Int,
    )

    private fun decodeEnvelopeHeader(envelope: ByteArray): EnvelopeHeader {
        val lengthPrefixBytes = VAL_4
        if (envelope.size < lengthPrefixBytes + FIXED_HEADER_BYTE_COUNT) {
            throw CodecError.EnvelopeTooShort
        }
        val buffer = ByteBuffer.wrap(envelope).order(ByteOrder.BIG_ENDIAN)
        val totalPayloadCount = buffer.int
        val envelopeBoundsError: CodecError? =
            when {
                totalPayloadCount > maxPayloadBytes ->
                    CodecError.PayloadTooLarge(totalPayloadCount, maxPayloadBytes)
                envelope.size < lengthPrefixBytes + totalPayloadCount -> CodecError.HeaderTruncated
                else -> null
            }
        if (envelopeBoundsError != null) throw envelopeBoundsError
        return EnvelopeHeader(buffer = buffer, totalEnvelopeBytes = lengthPrefixBytes + totalPayloadCount)
    }

    private fun validateMagicAndVersion(buffer: ByteBuffer) {
        val magic = ByteArray(MAGIC.size)
        buffer.get(magic)
        if (!magic.contentEquals(MAGIC)) throw CodecError.InvalidMagic
        val version = buffer.get()
        if (version != VERSION) throw CodecError.UnsupportedVersion(version)
    }

    private fun validatePayloadBounds(
        metadataLength: Int,
        payloadLength: Int,
        buffer: ByteBuffer,
        totalEnvelopeBytes: Int,
    ) {
        val payloadEnd = buffer.position() + metadataLength + payloadLength
        val hasInvalidPayloadBounds =
            metadataLength < 0 || payloadLength < 0 || payloadEnd > totalEnvelopeBytes
        if (hasInvalidPayloadBounds) {
            throw CodecError.HeaderTruncated
        }
    }

    companion object {
        val MAGIC = byteArrayOf(0x4F, 0x42, 0x4D, 0x46, 0x56, 0x32) // OBMFV2
        const val VERSION: Byte = 2
        const val FIXED_HEADER_BYTE_COUNT: Int = 34
        const val DEFAULT_MAX_PAYLOAD_BYTES: Int = 2 * 1024 * 1024

        fun isEncodedEnvelope(envelope: ByteArray): Boolean {
            val headerStart = VAL_4
            if (envelope.size < headerStart + MAGIC.size) return false
            return MAGIC.indices.all { index -> envelope[headerStart + index] == MAGIC[index] }
        }
    }
}
